jwt.verify.publickey.location property should support HTTP

**Describe the bug**

According to MP JWT spec `mp.jwt.verify.publickey.location` property should support HTTP/S protocol: https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/configuration.asciidoc#http-url-scheme

Then I published public key at https://gist.githubusercontent.com/lordofthejars/9a0b47298c807739a9eb1adfba50795e/raw/5f8c5b7a9c11ed455f4910ed2032d79612b5e6f6/quarkus.jwt.pub.pem

With valid token: https://gist.githubusercontent.com/lordofthejars/9a0b47298c807739a9eb1adfba50795e/raw/5f8c5b7a9c11ed455f4910ed2032d79612b5e6f6/quarkus.jwt.token

Then with the next configuration file:

```
mp.jwt.verify.publickey.location=https://gist.githubusercontent.com/lordofthejars/9a0b47298c807739a9eb1adfba50795e/raw/5f8c5b7a9c11ed455f4910ed2032d79612b5e6f6/quarkus.jwt.pub.pem
mp.jwt.verify.issuer=https://quarkus.io/using-jwt-rbac
```

And endpoint:

```
   @Claim(standard = Claims.preferred_username)
    String username;

    @GET
    @Path("/claim")
    public String getClaim() {
        return username;
    }
```

When I do the curl to the endpoint, the next exception is thrown:

```
JWT processing failed. Additional details: [[17] Unable to process JOSE object (cause: org.jose4j.lang.UnresolvableKeyException: Failed to resolve a key from: https://gist.githubusercontent.com/lordofthejars/9a0b47298c807739a9eb1adfba50795e/raw/5f8c5b7a9c11ed455f4910ed2032d79612b5e6f6/quarkus.jwt.pub.pem): JsonWebSignature{"kid":"\/privateKey.pem","typ":"JWT","alg":"RS256"}->eyJraWQiOiJcL
```

Then if I take the same public key and I copy it manually to disk and I refer to disk (META-INF/resources), then everything works as expected.

Then I set logging at TRACE level and I noticed this:

```
2019-10-10 12:03:28,167 DEBUG [io.sma.jwt.aut.pri.KeyLocationResolver] (executor-thread-1) Failed to create a key from the HTTPS JWK(S): org.jose4j.lang.JoseException: Parsing error: org.jose4j.json.internal.json_simple.parser.ParseException: Unexpected character (-) at position 0.
	at org.jose4j.json.JsonUtil.parseJson(JsonUtil.java:66)
	at org.jose4j.jwk.JsonWebKeySet.<init>(JsonWebKeySet.java:38)
	at org.jose4j.jwk.HttpsJwks.refresh(HttpsJwks.java:205)
	at org.jose4j.jwk.HttpsJwks.getJsonWebKeys(HttpsJwks.java:161)
	at io.smallrye.jwt.auth.principal.KeyLocationResolver.getHttpsJwk(KeyLocationResolver.java:119)
	at io.smallrye.jwt.auth.principal.KeyLocationResolver.tryAsJwk(KeyLocationResolver.java:107)
	at io.smallrye.jwt.auth.principal.KeyLocationResolver.resolveKey(KeyLocationResolver.java:95)
	at org.jose4j.jwt.consumer.JwtConsumer.processContext(JwtConsumer.java:213)
	at org.jose4j.jwt.consumer.JwtConsumer.process(JwtConsumer.java:433)
	at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.parse(DefaultJWTTokenParser.java:88)
	at io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator.validateClaimsSet(MpJwtValidator.java:41)
	at io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator.validate(MpJwtValidator.java:35)
	at io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator_ClientProxy.validate(MpJwtValidator_ClientProxy.zig:193)
	at org.wildfly.security.auth.realm.token.TokenSecurityRealm$TokenRealmIdentity.validateToken(TokenSecurityRealm
```

So when there is a URL we are trying to resolve the key as a JSON and not as a pem.

**Expected behavior**

We should offer similar behaviour when the file is local or online.

**Actual behavior**

Takes as JSON.


**Environment (please complete the following information):**
 - Quarkus 0.32.2
 - Mac OS Java 11
 - Quarkus in JVM mode.

@lordofthejars  yes I used and it shows this error:

```
Caused by: io.smallrye.jwt.auth.principal.ParseException: Failed to verify token
        at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.parse(DefaultJWTTokenParser.java:115)
        at io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator.validateClaimsSet(MpJwtValidator.java:41)
        ... 51 more
Caused by: org.jose4j.lang.UnresolvableKeyException: Failed to load a key from: http://localhost:14145/jwks
        at io.smallrye.jwt.auth.principal.KeyLocationResolver.<init>(KeyLocationResolver.java:79)
        at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.getKeyResolver(DefaultJWTTokenParser.java:229)
        at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.parse(DefaultJWTTokenParser.java:74)
        ... 52 more
```

And you tried to access with curl to localhost:14145/jwks to validate that
it is returning a valid JWKS object

Missatge de Heni-ghodbane <notifications@github.com> del dia dv., 11 d’oct.
2019 a les 12:28:

> @lordofthejars <https://github.com/lordofthejars> yes I used and it shows
> this error:
>
> Caused by: io.smallrye.jwt.auth.principal.ParseException: Failed to verify token
>         at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.parse(DefaultJWTTokenParser.java:115)
>         at io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator.validateClaimsSet(MpJwtValidator.java:41)
>         ... 51 more
> Caused by: org.jose4j.lang.UnresolvableKeyException: Failed to load a key from: http://localhost:14145/jwks
>         at io.smallrye.jwt.auth.principal.KeyLocationResolver.<init>(KeyLocationResolver.java:79)
>         at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.getKeyResolver(DefaultJWTTokenParser.java:229)
>         at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.parse(DefaultJWTTokenParser.java:74)
>         ... 52 more
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/issues/4493?email_source=notifications&email_token=AALSMYJGPGCWHOJZ2BRTB63QOBINFA5CNFSM4I7KYQSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEA7SNKQ#issuecomment-541009578>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALSMYME6B5XRCDJUHEV453QOBINFANCNFSM4I7KYQSA>
> .
>

--- a/bom/runtime/pom.xml
+++ b/bom/runtime/pom.xml
@@ -37,7 +37,7 @@
         <smallrye-open-api.version>1.1.19</smallrye-open-api.version>
         <smallrye-opentracing.version>1.3.2</smallrye-opentracing.version>
         <smallrye-fault-tolerance.version>2.1.2</smallrye-fault-tolerance.version>
-        <smallrye-jwt.version>2.0.9</smallrye-jwt.version>
+        <smallrye-jwt.version>2.0.10</smallrye-jwt.version>
         <smallrye-context-propagation.version>1.0.11</smallrye-context-propagation.version>
         <smallrye-reactive-streams-operators.version>1.0.10</smallrye-reactive-streams-operators.version>
         <smallrye-converter-api.version>1.0.10</smallrye-converter-api.version>

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcJsonWebTokenProducer.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcJsonWebTokenProducer.java
@@ -1,7 +1,5 @@
 package io.quarkus.oidc.runtime;
 
-import java.util.Set;
-
 import javax.annotation.Priority;
 import javax.enterprise.context.RequestScoped;
 import javax.enterprise.inject.Alternative;
@@ -16,6 +14,7 @@ import org.jose4j.jwt.consumer.JwtConsumerBuilder;
 import io.quarkus.oidc.IdToken;
 import io.quarkus.security.credential.TokenCredential;
 import io.quarkus.security.identity.SecurityIdentity;
+import io.smallrye.jwt.auth.cdi.NullJsonWebToken;
 
 @Priority(2)
 @Alternative
@@ -77,23 +76,5 @@ public class OidcJsonWebTokenProducer {
             return new OidcJwtCallerPrincipal(jwtClaims);
         }
         throw new IllegalStateException("Current identity not associated with an access token");
-    }
-
-    private static class NullJsonWebToken implements JsonWebToken {
-
-        @Override
-        public String getName() {
-            return null;
-        }
-
-        @Override
-        public Set<String> getClaimNames() {
-            return null;
-        }
-
-        @Override
-        public <T> T getClaim(String claimName) {
-            return null;
-        }
     }
 }

--- a/extensions/smallrye-jwt/runtime/src/main/java/io/quarkus/smallrye/jwt/runtime/auth/JwtPrincipalProducer.java
+++ b/extensions/smallrye-jwt/runtime/src/main/java/io/quarkus/smallrye/jwt/runtime/auth/JwtPrincipalProducer.java
@@ -1,7 +1,5 @@
 package io.quarkus.smallrye.jwt.runtime.auth;
 
-import java.util.Set;
-
 import javax.annotation.Priority;
 import javax.enterprise.context.RequestScoped;
 import javax.enterprise.inject.Alternative;
@@ -11,6 +9,7 @@ import javax.inject.Inject;
 import org.eclipse.microprofile.jwt.JsonWebToken;
 
 import io.quarkus.security.identity.SecurityIdentity;
+import io.smallrye.jwt.auth.cdi.NullJsonWebToken;
 
 @Priority(1)
 @Alternative
@@ -35,23 +34,5 @@ public class JwtPrincipalProducer {
             return (JsonWebToken) identity.getPrincipal();
         }
         throw new IllegalStateException("Current principal " + identity.getPrincipal() + " is not a JSON web token");
-    }
-
-    private static class NullJsonWebToken implements JsonWebToken {
-
-        @Override
-        public String getName() {
-            return null;
-        }
-
-        @Override
-        public Set<String> getClaimNames() {
-            return null;
-        }
-
-        @Override
-        public <T> T getClaim(String claimName) {
-            return null;
-        }
     }
 }

